pkgs/os-specific/darwin/apple-source-releases/AvailabilityVersions/0001-Support-setting-an-upper-bound-on-versions.patch

```diff
@@ -0,0 +1,249 @@
+From a5f693f96e5db84dbe46a4757b1d845a7fc3f2c7 Mon Sep 17 00:00:00 2001
+From: Randy Eckenrode <randy@largeandhighquality.com>
+Date: Tue, 2 Jul 2024 20:04:56 -0400
+Subject: [PATCH] Support setting an upper bound on versions
+
+---
+ availability | 99 ++++++++++++++++++++++++++++++++++------------------
+ 1 file changed, 66 insertions(+), 33 deletions(-)
+
+diff --git a/availability b/availability
+index 8ebd250..fd25fd0 100755
+--- a/availability
++++ b/availability
+@@ -17,12 +17,39 @@ MIN_PYTHON = (3, 7) #Required for ordered dictionaries as default
+ if sys.version_info < MIN_PYTHON:
+     sys.exit("Python %s.%s or later is required.\n" % MIN_PYTHON)
+ 
++
++def parse_version(ver):
++    if hasattr(ver, "string"):
++        ver = ver.string()
++
++    parts = ver.split(".")
++    if len(parts) == 1:
++        return int(parts[0]), 0, 0
++    elif len(parts) == 2:
++        return int(parts[0]), int(parts[1]), 0
++    else:
++        return tuple(map(int, parts))
++
++def version_older_or_equal(lhs, rhs):
++    if not rhs:
++        return True
++
++    lhs_major, lhs_minor, lhs_patch = parse_version(lhs)
++    rhs_major, rhs_minor, rhs_patch = parse_version(rhs)
++
++    return (
++        lhs_major < rhs_major
++        or (lhs_major == rhs_major and lhs_minor < rhs_minor)
++        or (lhs_major == rhs_major and lhs_minor == rhs_minor and lhs_patch <= rhs_patch)
++    )  
++     
++
+ # The build script will embed the DSL content here, otherwise we build it at runtime
+ dslContent = None
+ # @@INSERT_DSL_CONTENT()@@
+ 
+ class VersionSetDSL:
+-    def __init__(self, data):   self.parsedDSL = self.Parser(data)
++    def __init__(self, data, threshold):   self.parsedDSL = self.Parser(data, threshold)
+     def sets(self):             return self.parsedDSL.version_sets
+     def platforms(self):        return self.parsedDSL.platforms
+ 
+@@ -104,12 +131,15 @@ class VersionSetDSL:
+                 self.availability_deprecation_define_name = optionals["availability_deprecation_define_name"]
+             if "version_define_name" in optionals:
+                 self.availability_define_prefix = f"__{optionals['version_define_name']}_"
+-        def add_version(self, version): return self.versions.append(version);
++        def add_version(self, version, threshold):
++            if version_older_or_equal(version, threshold):
++                return self.versions.append(version);
```

Comment:
is this missing a return when the expr in the if evaluates to false?

Reply:
Yes, that’s intentional. It’s only updating the list of versions if they pass the filter.

Reply:
I dropped the `return` and `;`.  The [original code][1] has them, but `list.append` returns `None` in Python, so there’s no reason to `return` (especially if it causes confusion about the function’s return type).

[1]: https://github.com/apple-oss-distributions/AvailabilityVersions/blob/8f08508d9013bcc371e6c47d8bead2a9d5fdea07/availability#L107